### PR TITLE
Simplify and fix the Int128 *, /, and % logic

### LINKED ISSUE #75416: bugs: System.Int128 operator * and % at overflow
### In case of an overflow for multiplication `*` and remainder `%`  `System.Int128` gives wrong results.
All other C# integer types behave in this case correctly like standard on all machines and platforms.
The problem is that there are many algorithms that assume correct partial results in case of an overflow as part of the calculation.
All these algorithms can fail with Int128 and it's dangerous because the user wouldn't expect it.

The bug is in `public static Int128 operator *(Int128 left, Int128 right)`
Easy to see that the case is simply not handled and the final `return new Int128(result.Upper | sign, result.Lower);` copys the sign always. 
This is like a proof as the sign of the result shouldn't be affected in this case.
It also follows that the error occurs in large number ranges, in other ranges not and in sub-ranges alternating.

Exactly the same issue in `public static Int128 operator %(Int128 left, Int128 right)`.

As reference, to reproduce and test the following code can be used: [Test System.Int128 functions, operator and INumber implementation](https://github.com/c-ohle/RationalNumerics/blob/5cdc322b5d06be62718e1f653bc41afe678ab2b1/Test/Program.cs#L496).
A template solution for arbitrary number type for NET7 is used, which of course also supports int128. 
An correct implementation for multiplication and remainder can be taken from there.
The tests also show that all other `System.Int128` functions and operators handle the overflows correctly. 

### Reproduction Steps

```csharp
var a = (Int128.MaxValue - 10) * +100; // -1100 ok
var b = (Int128.MaxValue - 10) * -100; // -170141183460469231731687303715884104628 wrong, should be +1100
```

### Expected behavior

```csharp
var a = (Int128.MaxValue - 10) * +100; // -1100 
var b = (Int128.MaxValue - 10) * -100; // +1100

```

### Actual behavior

```csharp
var a = (Int128.MaxValue - 10) * +100; // -1100 ok
var b = (Int128.MaxValue - 10) * -100; // -170141183460469231731687303715884104628 wrong, should be +1100
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves #75416

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 1]
```diff
@@ -1273,76 +1241,38 @@ public static Int128 Log2(Int128 value)
...
-            ulong sign = (left._upper ^ right._upper) & (1UL << 63);
-
-            if (IsNegative(left))
-            {
-                left = ~left + 1U;
-            }
+            Int128 upper = BigMul(left, right, out Int128 lower);
 
-            if (IsNegative(right))
+            if (((upper != 0) || (lower < 0)) && (~upper != 0))
```
Comment (tfenise) [inline src/libraries/System.Private.CoreLib/src/System/Int128.cs:1253]: What if `upper == -1` and `lower >= 0`?
```suggestion
            if (((upper != 0) || (lower < 0)) && ((~upper != 0) || (lower >= 0)))
```

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 2]
```diff
@@ -1273,76 +1241,38 @@ public static Int128 Log2(Int128 value)
...
-                result.Upper | sign,
-                result.Lower
-            );
+        internal static Int128 BigMul(Int128 left, Int128 right, out Int128 lower)
+        {
+            // This follows the same logic as is used in `long Math.BigMul(long, long, out long)`
+
+            UInt128 upper = UInt128.BigMul((UInt128)(left), (UInt128)(right), out UInt128 ulower);
+            lower = (Int128)(ulower);
+            return (Int128)(upper) - ((left >> 63) & right) - ((right >> 63) & left);
```
Comment (tfenise) [inline src/libraries/System.Private.CoreLib/src/System/Int128.cs:1275]: 63?
```suggestion
            return (Int128)(upper) - ((left >> 127) & right) - ((right >> 127) & left);
```

**Review by tfenise — CHANGES_REQUESTED** [review]
I think there are some bugs.

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Int128.cs]: Ah, yep. Had this locally but forgot to commit it.

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs (hunk 2 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Int128.cs]: Copy/paste error, thanks for the catch!

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 3]
```diff
@@ -1273,76 +1241,38 @@ public static Int128 Log2(Int128 value)
...
             {
-                right = ~right + 1U;
-            }
+                // The upper bits can safely be either Zero or AllBitsSet
+                // where the former represents a positive value and the
+                // latter a negative value.
+                //
+                // However, when the upper bits are Zero, we also need to
+                // confirm the lower bits are positive, otherwise we have
+                // a positive value greater than MaxValue and should throw
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/Int128.cs:1261]: This comment should be updated to match the user-suggested fix

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 4]
```diff
@@ -1273,76 +1241,38 @@ public static Int128 Log2(Int128 value)
...
-            ulong sign = (left._upper ^ right._upper) & (1UL << 63);
-
-            if (IsNegative(left))
-            {
-                left = ~left + 1U;
-            }
+            Int128 upper = BigMul(left, right, out Int128 lower);
 
-            if (IsNegative(right))
+            if (((upper != 0) || (lower < 0)) && ((~upper != 0) || (lower >= 0)))
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/Int128.cs:1253]: Something about this logic is hard to parse but I can't figure out a better way to structure it.

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 5]
```diff
@@ -1172,17 +1172,13 @@ public static Int128 Log2(Int128 value)
 
             UInt128 result = (UInt128)(left) / (UInt128)(right);
 
-            if (result == 0U)
-            {
-                sign = 0;
-            }
-            else if (sign != 0)
+            if (sign != 0)
```
[the remark this answers is not in the document]
Comment (dakersnar) [reply src/libraries/System.Private.CoreLib/src/System/Int128.cs:1175]: Additionally, the comment above it should say "two's complement"

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 6]
```diff
@@ -1273,76 +1241,38 @@ public static Int128 Log2(Int128 value)
...
-            {
-                result = ~result + 1U;
-            }
-
-            return new Int128(
-                result.Upper | sign,
-                result.Lower
-            );
+            // Multiplication is the same for signed and unsigned provided the "upper" bits aren't needed
+            return (Int128)((UInt128)(left) * (UInt128)(right));
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/Int128.cs:1245]: Is this a logic change or a simplification?

**Review by dakersnar — APPROVED** [review]
Added some comments, once those are addressed LGTM

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Int128.cs:1175]: Its actually more efficient to not make it a bool since we can avoid the normalization and make it just a `test` here instead.

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs (hunk 6 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Int128.cs:1245]: Both, and a regression test was added.

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 7]
```diff
@@ -1227,36 +1223,8 @@ public static Int128 Log2(Int128 value)
...
-            {
-                result = ~result + 1U;
-            }
-
-            return new Int128(
-                result.Upper | sign,
-                result.Lower
-            );
+            Int128 quotient = left / right;
+            return left - (quotient * right);
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Int128.cs:1227]: Both, although no error was provided in the original issue and I've not worked out a repro case yet.

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs (hunk 3 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Int128.cs:1261]: Fixed.

**Comment by tannergooding** [comment]
Logged https://github.com/dotnet/runtime/issues/75566 for the Mono llvmaot failure, which is unrelated.
